qubes/storage/zfs.py

```diff
@@ -0,0 +1,1219 @@
+                        this may be incorrect and should probably be checked.
+                        TODO.
+        """
+
+        self.vm_name = vm_name
+        self.name = name
+        assert self.name
+        self.pool = pool # <- ref to parent ZFSQpool
+        self.log = logging.getLogger(
+            "qubes.storage.zfs.ZFSQzvol.{}".format(zfs_ns))
+        self.log.warning('ZFSQzvol kwargs {}'.format(kwargs))
+        self.zfs_ns = zfs_ns # prefix to use before anything
+        assert self.zfs_ns
+        if self.zfs_ns.startswith("-"):
+            raise qubes.storage.StoragePoolException(
+                "Invalid zfs_ns; prefixed with '-'")
+
+        # TODO verify/update this by retrieving ?
+        # if lzc_exists(whatever):
+        #   self._size = lzc_get_props(whatever, "volsize")
+        self._size = size
+
+        self.vid = "/".join([self.zfs_ns,
+                             "vm",
+                             self.vm_name,
+                             self.name])
+        self.import_path = b"/".join([self.zfs_ns.encode(), b"import",
+                                      self.vm_name.encode()])
+        self._vid_import = b"/".join([self.import_path, self.name.encode()])
+
+        self.save_on_stop = False
+        self.snap_on_start = False
+        self.rw = rw
+        super().__init__(name=self.name,
+                         pool=self.pool,
+                         vid=self.vid,
+                         rw=self.rw,
+                         size=self._size)
+
+        self.log.warning("dataset_path assert {!r} {!r} {!r}".format(self.zfs_ns, b"vm", self.vm_name))
+        self.dataset_path = b"/".join(
+            [
+                self.zfs_ns.encode(),
+                b"vm",
+                self.vm_name.encode(),
+            ]
+        )
+        # sanity check:
+        assert self.dataset_path + b"/" + self.name.encode() == self.vid.encode()
+        self._vid_snap = self.vid.encode() + b"@latest"
+
+
+        if self.save_on_stop:
+            raise qubes.storage.StoragePoolException(
+                "TODO save_on_stop not supported"
+            )
+
+        self._size = size
+        self._lock = asyncio.Lock()
+
+    @locked
+    @asyncio.coroutine
+    def create(self):
+        assert self.vid
+        assert self.size
+
+        self.log.warning("zfs_zvol:create() vid {}".format(self.vid))
+
+        # if vm-specific datasets doesn't exist, create them:
+        for namespace in [self.import_path, self.dataset_path]:
+            try:
+                make_dataset = not libzfs_core.lzc_exists(namespace)
+            except libzfs_core.exceptions.ParentNotFound:
+                make_dataset = True
+            finally:
+                if make_dataset:
+                    self.log.warning(
+                        "zfs - creating namespace dataset {}".format(
+                            namespace
+                        )
+                    )
+                    libzfs_core.lzc_create(
```

Comment:
No, it blocks for a little bit of I/O, but under normal circumstances it's instantaneous.
It might block if you're doing this on top of NFS or whatever, but I think we're okay doing this in-thread.